File: tmp_metadata_modifier/tauto_modify/actions.py
# Copyright 2023 The ChromiumOS Authors
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""Functions which return "actions" that modify a given ControlFile object."""

# Each action should take in a single ControlFile and return a boolean value
# of whether the control file was modified or not.

def remove_contacts(emails):
    """Return an action which removes the given list of emails from 'contacts'.

    Args:
        emails: A list of string emails, e.g. ['foo@google.com'].

    Returns:
        An action function that acts on a ControlFile and returns a boolean.
    """
    def output(cf):
        if 'contacts' not in cf.metadata:
            return False
        modified = False
        for email in emails:
            if email in cf.metadata['contacts']:
                cf.metadata['contacts'].remove(email)
                modified = True
        return modified
    return output


def append_contacts(emails):
    """Return an action which appends the given emails to 'contacts'.

    Args:
        emails: A list of string emails, e.g. ['foo@google.com'].

    Returns:
        An action function that acts on a ControlFile and returns a boolean.
    """
    def output(cf):
        if 'contacts' not in cf.metadata:
            return False
        for email in emails:
            if email in cf.metadata['contacts']:
                cf.metadata['contacts'].remove(email)
        cf.metadata['contacts'] += emails
        return True
    return output


def prepend_contacts(emails):
    """Return an action which prepends the given emails to 'contacts'.

    Args:
        emails: A list of string emails, e.g. ['foo@google.com'].

    Returns:
        An action function that acts on a ControlFile and returns a boolean.
    """
    def output(cf):
        if 'contacts' not in cf.metadata:
            return False
        for email in emails:
            if email in cf.metadata['contacts']:
                cf.metadata['contacts'].remove(email)
        cf.metadata['contacts'] = emails + cf.metadata['contacts']
        return True
    return output


def set_hw_agnostic():
    """Return an action which sets 'hw_agnostic' to be true.

    Returns:
        An action function that acts on a ControlFile and returns a boolean.
    """
    def output(cf):
        if 'hw_agnostic' not in cf.metadata or cf.metadata['hw_agnostic'] != True:
            cf.metadata['hw_agnostic'] = True
            return True
        return False
    return output


def unset_hw_agnostic():
    """Return an action which removes 'hw_agnostic' if it is present.

    Returns:
        An action function that acts on a ControlFile and returns a boolean.
    """
    def output(cf):
        if 'hw_agnostic' in cf.metadata:
            del cf.metadata['hw_agnostic']
            return True
        return False
    return output

File: tmp_metadata_modifier/tast_modify/actions.go
// Copyright 2023 The ChromiumOS Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package main

import ()

// Action is a function which modifies a test file.
// It is applied to a TestFile object and returns true if the file was modified
// as a result of the action (or an error if something goes wrong).
type Action func(*TestFile) (bool, error)

// SetContactsAction replaces the entire contacts for a test with
// the given values.
func SetContactsAction(emails []string) Action {
	return func(f *TestFile) (bool, error) {
		f.SetContacts(emails)
		return true, nil
	}
}

// RemoveContactsAction returns an action that removes the given
// contact emails from the test declaration.
func RemoveContactsAction(emails []string) Action {
	return func(f *TestFile) (bool, error) {
		modified := false
		for _, email := range emails {
			modified = modified || f.RemoveContact(email)
		}
		return modified, nil
	}
}

// ReplaceContactAction returns an action that replaces the given oldEmail
// with the given newEmail in the test's contacts.
func ReplaceContactAction(oldEmail, newEmail string) Action {
	return func(f *TestFile) (bool, error) {
		contacts := f.Contacts()
		for i, elt := range contacts {
			if elt == oldEmail {
				contacts[i] = newEmail
				f.SetContacts(contacts)
				return true, nil
			}
		}
		return false, nil
	}
}

// AppendContactsAction returns an action that appends the given emails to
// a test's contacts, deleting them elsewhere in the list if already present.
func AppendContactsAction(emails []string) Action {
	return func(f *TestFile) (bool, error) {
		for _, email := range emails {
			_ = f.RemoveContact(email)
		}
		contacts := append(f.Contacts(), emails...)
		f.SetContacts(contacts)
		return true, nil
	}
}

// PrependContactsAction returns an action that prepends the given emails to
// a test's contacts, deleting them elsewhere in the list if already present.
func PrependContactsAction(emails []string) Action {
	return func(f *TestFile) (bool, error) {
		for _, email := range emails {
			_ = f.RemoveContact(email)
		}
		contacts := append(emails, f.Contacts()...)
		f.SetContacts(contacts)
		return true, nil
	}
}
